relayer/middleware/metrics.middleware.ts

```diff
@@ -1,45 +1,173 @@
+import { RelayJob, StorageContext } from "../storage/storage";
 import { Counter, Registry } from "prom-client";
 
-interface MetricsOpts {
+type MetricRecord = Record<string, string | number>;
+
+class MeasuredRelayJob {
+  private job?: RelayJob;
+  private startTime: number;
+  private endTime: number;
+
+  constructor(startTime: number, endTime: number, job?: RelayJob) {
+    this.job = job;
+    this.startTime = startTime;
+    this.endTime = endTime;
+  }
+
+  private currentAttempts(): number {
+    return this.job?.attempts ?? 0;
+  }
+
+  private maxAttempts(): number {
+    return this.job?.maxAttempts ?? Number.MAX_SAFE_INTEGER;
+  }
+
+  hasReachedMaxAttempts(): boolean {
+    return this.currentAttempts() >= this.maxAttempts();
+  }
+
+  vaaTimestamp(): number {
+    // vaa.timestamp is in seconds
+    return this.job?.data?.parsedVaa?.timestamp * 1000 ?? 0;
+  }
+
+  processingTime(): number {
+    return this.endTime - this.startTime;
+  }
+
+  totalTime(): number {
+    if (this.vaaTimestamp() > 0) {
+      return this.endTime - this.vaaTimestamp();
+    }
+
+    return 0;
+  }
+}
+
+const status = "status";
+const terminal = "terminal";
+
+export interface MetricsOpts<C extends StorageContext> {
   registry?: Registry;
+  labels?: MetricLabelsOpts<C>;
+  processingTimeBuckets?: number[];
+}
+
+export class MetricLabelsOpts<C extends StorageContext> {
+  labelNames: string[];
+  customizer: (context: C) => Promise<MetricRecord>;
+
+  constructor(
+    labelNames: string[] = [],
+    customizer: (ctx: C) => Promise<MetricRecord>,
+  ) {
+    this.labelNames = labelNames;
+    this.customizer = customizer;
+  }
 }
 
-export function metrics(
-  opts: MetricsOpts = {},
-): Middleware<Context & { job?: Job }> {
+const getLabels = async <C extends StorageContext>(
+  ctx: C,
+  opts: MetricsOpts<C>,
+  failed: boolean,
+) => {
+  const defaultLabels = { status: failed ? "failed" : "succeeded" };
+
+  try {
+    const labels: MetricRecord = await (opts.labels?.customizer(ctx) ??
+      Promise.resolve({}));
+
+    const allowedKeys = opts.labels?.labelNames ?? [];
+    const validatedLabels: MetricRecord = {};
+    allowedKeys.forEach(key => {
+      if (labels[key]) {
+        validatedLabels[key] = labels[key];
+      }
+    });
+
+    return { ...defaultLabels, ...validatedLabels };
+  } catch (error) {
+    ctx.logger?.error(
+      "Failed to customize metric labels, please review",
+      error,
+    );
+
+    return defaultLabels;
+  }
+};
+
+export function metrics<C extends StorageContext>(
+  opts: MetricsOpts<C> = {},
+): Middleware<C> {
   opts.registry = opts.registry || register;
+
   const processedVaasTotal = new Counter({
     name: "vaas_processed_total",
-    help: "Number of vaas processed successfully or unsuccessfully.",
+    help: "Number of incoming vaas processed successfully or unsuccessfully.",
+    labelNames: [].concat(opts.labels?.labelNames ?? []),
```

Comment:
Looks like concat operation here isn’t necessary, PTAL
```suggestion
    labelNames: opts.labels?.labelNames ?? []
```